src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1895 @@
+                sd_bus_message *msg,
+                const Job *j,
+                sd_json_variant *json,
+                sd_bus_error *error) {
+
+        /* Called when job finishes w/ a successful exit code, but before any work begins.
+         * This happens when there is no candidate (i.e. we're already up-to-date), or
+         * specified update is already installed. */
+        return sd_bus_error_set_errno(error, -EALREADY);
+}
+
+static int target_method_update(sd_bus_message *msg, void *userdata, sd_bus_error *error) {
+        Target *t = ASSERT_PTR(userdata);
+        _cleanup_(job_freep) Job *j = NULL;
+        const char *version, *action;
+        int r;
+
+        assert(msg);
+
+        r = sd_bus_message_read(msg, "s", &version);
+        if (r < 0)
+                return r;
+
+        if (isempty(version))
+                action = "org.freedesktop.sysupdate1.update";
+        else
+                action = "org.freedesktop.sysupdate1.update-to-version";
+
+        const char *details[] = {
+                "class", target_class_to_string(t->class),
+                "name", t->name,
+                "version", version,
+                NULL
+        };
+
+        r = bus_verify_polkit_async(
+                msg,
+                action,
+                details,
+                &t->manager->polkit_registry,
+                error);
+        if (r < 0)
+                return r;
+        if (r == 0)
+                return 1; /* Will call us back */
+
+        r = job_new(JOB_UPDATE, t, msg, target_method_update_finished_early, &j);
+        if (r < 0)
+                return r;
+
+        j->detach_on_ready = true;
+
+        j->version = strdup(version);
+        if (!j->version)
+                return log_oom();
+
+        r = job_start(j);
+        if (r < 0)
+                return sd_bus_error_set_errnof(error, r, "Failed to start job: %m");
+        TAKE_PTR(j);
+
+        return 1;
+}
+
+static int target_method_vacuum_finish(
+                sd_bus_message *msg,
+                const Job *j,
+                sd_json_variant *json,
+                sd_bus_error *error) {
+
+        uint64_t instances;
+        
+        assert(json);
+
+        instances = sd_json_variant_unsigned(sd_json_variant_by_key(json, "removed"));
+        
+        return sd_bus_reply_method_return(msg, "u", instances);
+}
+
+static int target_method_vacuum(sd_bus_message *msg, void *userdata, sd_bus_error *error) {
+        Target *t = ASSERT_PTR(userdata);
+        _cleanup_(job_freep) Job *j = NULL;
+        int r;
+
+        assert(msg);
+
+        const char *details[] = {
+                "class", target_class_to_string(t->class),
+                "name", t->name,
+                NULL
+        };
+
+        r = bus_verify_polkit_async(
+                msg,
+                "org.freedesktop.sysupdate1.vacuum",
+                details,
+                &t->manager->polkit_registry,
+                error);
```

Comment:
doesn't matter much, but we typically want that function calls that are line broken use 16space indenting for the parameters, to visually distinguish them from regular indented code blocks.